### PR TITLE
[AUTO-MERGE] Merge main into develop via merge-main-to-develop

### PR DESCRIPTION
**Pull request by idanov**
A new change in main cannot be merged into develop as part of the regular sync job, hence this PR. Please resolve the conflicts manually, and make sure to obtain 2 approvals once the builds pass.

### IMPORTANT NOTICE

Please let CircleCI merge this PR automatically, with merge commit enabled.

### CODE REVIEW

**Comment by merelcht** [comment]
@noklam Can you help resolving the merge conflicts? I tried going through it, but this contains your tidying of `starter.py` and those changes are very hard to merge into develop..

**Comment by noklam** [comment]
https://github.com/kedro-org/kedro/pull/3280. It seems that the conflicts are repeating. It seems that it wasn't merged properly, it should be a merge commit but it get squashed?

**Comment by deepyaman** [comment]
> #3280. It seems that the conflicts are repeating. I suspect it wasn't merged properly, it should be a merge commit but it get squashed @deepyaman.
> 
> I think we should cherry-pick the commits from #3280

Sorry, yes, I set it to auto-squash-merge. I don't think I have permission to merge manually, and policy prevents merging with merge commit via the web interface, so I'll let somebody else do the honors (rather than my disservice :P).

File: docs/source/data/how_to_create_a_custom_dataset.md [hunk 1]
```diff
@@ -569,7 +569,7 @@ class ImageDataset(AbstractVersionedDataset):
     ...
 ```
 
-We provide additional examples of [how to use parameters through the data catalog's YAML API](./data_catalog_yaml_examples.md). For an example of how to use these parameters in your dataset's constructor, please see the [SparkDataset](/kedro_datasets.spark.SparkDataset)'s implementation.
+We provide additional examples of [how to use parameters through the data catalog's YAML API](./data_catalog_yaml_examples.md). For an example of how to use these parameters in your dataset's constructor, please see the implementation of the {class}`SparkDataset<kedro-datasets:kedro_datasets.spark.SparkDataset>`.
```
Comment (deepyaman) [inline docs/source/data/how_to_create_a_custom_dataset.md:572]:
```suggestion
We provide additional examples of [how to use parameters through the data catalog's YAML API](./data_catalog_yaml_examples.md). For an example of how to use these parameters in your dataset's constructor, see the implementation of the {class}`SparkDataset<kedro-datasets:kedro_datasets.spark.SparkDataset>`.
```

File: docs/source/data/how_to_create_a_custom_dataset.md [hunk 2]
```diff
@@ -511,7 +511,7 @@ Inspect the content of the data directory to find a new version of the data, wri
...
 
-There is one dataset that is an exception: [SparkDataset](/kedro_datasets.spark.SparkDataset). The explanation for this exception is that [Apache Spark](https://spark.apache.org/) uses its own parallelism and therefore doesn't work with Kedro [ParallelRunner](/kedro.runner.ParallelRunner). For parallelism within a Kedro project that leverages Spark please consider the alternative [ThreadRunner](/kedro.runner.ThreadRunner).
+There is one dataset that is an exception: {class}`SparkDataset<kedro-datasets:kedro_datasets.spark.SparkDataset>`. The explanation for this exception is that [Apache Spark](https://spark.apache.org/) uses its own parallelism and therefore doesn't work with Kedro [ParallelRunner](/kedro.runner.ParallelRunner). For parallelism within a Kedro project that leverages Spark please consider the alternative [ThreadRunner](/kedro.runner.ThreadRunner).
```
Comment (deepyaman) [inline docs/source/data/how_to_create_a_custom_dataset.md:514]:
```suggestion
There is one dataset that is an exception: {class}`SparkDataset<kedro-datasets:kedro_datasets.spark.SparkDataset>`. [Apache Spark](https://spark.apache.org/) uses its own parallelism and therefore doesn't work with Kedro [ParallelRunner](/kedro.runner.ParallelRunner). For parallelism within a Kedro project that leverages Spark please consider the alternative [ThreadRunner](/kedro.runner.ThreadRunner).
```
I'd question that you need a link to Spark, but I suppose that can help a completely new user who stumbles in here without knowing what Spark is?

**Comment by deepyaman** [comment]
@noklam In retrospect (after doing the resolution), it seems the only change at the time of doing this, in `main`, was https://github.com/kedro-org/kedro/pull/3294. That is reflected in the rather-minimal diff. Almost all of the changes then are to the kedro/framework/cli/starters.py file, and I've tried to entirely reflect your changes from #3294 by manually going through there. However, there has also been code added in `develop` for this file; maybe you can take a look to see if it makes sense based on the standard you used before (or can reorganize in a future PR). I did a bit of reorganization of the project creation constants, at least.

**Comment by noklam** [comment]
@deepyaman thank you! You don't need to merge it as there is a background job scanning the PR every hour to do the merge commit if it has enough approval. I will have a look at the starter.

**Comment by deepyaman** [comment]
> @deepyaman thank you! You don't need to merge it as there is a background job scanning the PR every hour to do the merge commit if it has enough approval. I will have a look at the starter.

Ah, right, I forgot. I should know this, since I replicated this on an internal repo. 😅

**Comment by noklam** [comment]
Thanks! I have moved stuff around, introspective I should have just refactor this in `develop` to avoid the conflicts

**Comment by merelcht** [comment]
I'm surprised there are only 2 files changed now, because this should also include some doc changes and your work on the `StopIteration` exception.

**Comment by noklam** [comment]
@merelcht  The change was merged in https://github.com/kedro-org/kedro/pull/3280 already. #3294 is the only new change as I understood.

In fact, the only change should be `starters.py` and I am going to update the PR

**Comment by deepyaman** [comment]
> @merelcht The change was merged in #3280 already. #3294 is the only new change as I understood.
> 
> In fact, the only change should be `starters.py` and I am going to update the PR

@noklam I intentionally deleted that; that was an old docstring for `__init__` method which is gone; the class already has a docstring. ;)

**Review by deepyaman — APPROVED** [review]

**Comment by deepyaman** [comment]
> I'm surprised there are only 2 files changed now, because this should also include some doc changes and your work on the `StopIteration` exception.

Yeah, the changes are already there; it's more about adding back the commit history (not included because of previous squash-merge).

**Review by merelcht — APPROVED** [review]

**Pull request merged**